core/amber/src/main/python/core/util/customized_queue/linked_blocking_multi_queue.py

```diff
@@ -0,0 +1,436 @@
+
+        # thread-safe in CPython
+        self.priority_groups: List[LinkedBlockingMultiQueue.PriorityGroup] = list()
+        self.sub_queue_selection = LinkedBlockingMultiQueue.DefaultSubQueueSelection(
+            self.priority_groups
+        )
+
+    def put(self, key: str, item: T) -> None:
+        """
+        Put one item into the SubQueue specified by the key.
+
+        :param key: the identifier of a SubQueue.
+        :param item: Any instance.
+        :raises KeyError for non-existing keys.
+        :return: None
+        """
+        self._get_sub_queue(key).put(item)
+
+    def get(self) -> T:
+        """
+        Blocking get the next available item from the queue.
+        - Disabled SubQueues are considered empty and will not be fetched.
+        - When multiple SubQueues are enabled and have items, it selects the SubQueue
+        by the order specified by the self.sub_queue_selection strategy.
+
+        :return: T, Any item that is available to the fetched.
+        """
+        self.take_lock.acquire()
+        try:
+            while self.total_count.value == 0:
+                self.not_empty.wait()
+
+            # at this point we know there is an element
+            sub_queue = self.sub_queue_selection.get_next()
+            item = sub_queue.dequeue()
+            sub_queue.count.dec()
+            if self.total_count.get_and_dec() > 1:
+                # sub queue still has element
+                self.not_empty.notify()
+        finally:
+            self.take_lock.release()
+
+        return item
+
+    def peek(self) -> Optional[T]:
+        """
+        Peek the next available item from the queue.
+        - When no item is available, it returns None.
+        - Otherwise, it acts the same as LinkedBlockingMultiQueue.get() but
+        without actually taking the item out from the queue.
+
+        :return: Optional[T], could be the available item or None.
+        """
+        self.take_lock.acquire()
+        try:
+            if self.total_count.value == 0:
+                return None
+            else:
+                return self.sub_queue_selection.peek()
+        finally:
+            self.take_lock.release()
+
+    def enable(self, key: str) -> None:
+        """
+        Enables a SubQueue, specified by key. This action acquires all locks.
+
+        :param key: the identifier of the SubQueue.
+        :raises KeyError for non-existing keys.
+        :return: None
+        """
+        self._get_sub_queue(key).enable()
+
+    def disable(self, key: str) -> None:
+        """
+        Disables a SubQueue, specified by key. This action acquires all locks.
+
+        :param key: the identifier of the SubQueue.
+        :raises KeyError for non-existing keys.
+        :return: None
+        """
+        self._get_sub_queue(key).disable()
+
+    def size(self, key: Optional[str] = None) -> int:
+        """
+        Get the current number of elements of the queue, or a specific SubQueue if
```

Comment:
I suggest we document it clearly as the total size of all sub queues in this queue

Reply:
done.